apps/emqx_bridge_snowflake/README.md

```diff
@@ -0,0 +1,22 @@
+# EMQX Snowflake Bridge
+
+This application provides connector and action implementations for the EMQX to integrate with Snowflake as part of the EMQX data integration pipelines.
+Users can leverage [EMQX Rule Engine](https://docs.emqx.com/en/enterprise/latest/data-integration/rules.html) to create rules that publish message data to Snowflake.
+
+## Documentation
+
+Refer to [Rules engine](https://docs.emqx.com/en/enterprise/latest/data-integration/rules.html) for the EMQX rules engine introduction.
+
+## Implementation details
+
+At the time of writing, it's not possible to use Snowflake's SDKs which support [Snowpipe Streaming](https://github.com/snowflakedb/snowflake-ingest-java/blob/master/README.md#snowpipe-streaming).  Here, we try to approximate what the SDK does, which is to buffer rows into (Parquet) files before uploading (staging) and ingesting them in bulk using currently undocumented Snowpipe APIs.
+
+In Snowflake, files are staged either in internal or external stages (AWS, Azure, GCP).  Internal stages must execute [`PUT`](https://docs.snowflake.com/en/user-guide/data-load-local-file-system-stage) statements using the ODBC API.  We buffer messages/rows using [`emqx_connector_aggregator`](../emqx_connector_aggregator) in container files (currently, only CSV containers are supported) until either a time or a record count threshold is hit.  One or more temporary files might be created locally, depending on the volume of data and configured maximum file limits.  These files are then uploaded via an ODBC connection using a `PUT file://<local path> @<stage path>` command.  Once all files are staged, we load the data using [Snowpipe's `insertFiles` REST API](https://docs.snowflake.com/en/user-guide/data-load-snowpipe-rest-load).  There's no equivalent ODBC API for this ingestion at this time, except directly specifying `COPY INTO` statements, which require allocating and specifying an [warehouse](https://docs.snowflake.com/en/user-guide/warehouses-overview).  After Snowpipe ingests the data, it'll be available in the table defined in the pipe's DDL `COPY INTO` statement.
```

Comment:
the format is Parquet, but here only CSV.
please help to connect the dots.

Reply:
could you expand a bit on the question?  I'm not I follow.

Reply:
I think this bit is confusing

> we try to approximate what the SDK does, which is to buffer rows into (Parquet) files before uploading

You meant buffering into Parquet as in what SDK does, but one could also read it as "while trying to approximate what SDK does, we buffer into Parquet". And below you talk about buffering into CSV.

Reply:
Btw, saying that we try to approximate the Streaming SDK is not entirely correct, since it's using completely different API endpoints, and provides very low latency.

Reply:
We are actually implementing a robust integration for Snowpipe API, but not for Snowpipe Streaming.

Reply:
> You meant buffering into Parquet as in what SDK does, but one could also read it as "while trying to approximate what SDK does, we buffer into Parquet". And below you talk about buffering into CSV.

Aha, now I see the confusion, thanks for the explanation.  

What I tried to convey was these separate statements: i) the SDK works by buffering into a particular file format and calling undocumented APIs; ii) we try to approximate the streaming behavior using official, documented APIs and also using buffering.  In the following paragraph, I detail our implementation, but it was not the intention to say we exactly mimic the SDK.

I'll try to rephrase this section.

> Btw, saying that we try to approximate the Streaming SDK is not entirely correct, since it's using completely different API endpoints, and provides very low latency.

> We are actually implementing a robust integration for Snowpipe API, but not for Snowpipe Streaming.

Good point.  The intention was to approximate streaming based on what we saw the SDK do, which does involve buffering to a file before essentially staging and inserting the files, but you make a good point: their hidden Snowpipe Streaming APIs most likely have different latency guarantees than the official route.  I've see a single CSV with 3 rows take ~ 10 seconds or more to be ingested, so it's probably more batch-oriented and asynchronous than the SDK channel APIs.